### ISSUE TITLE
Ability to deploy from build

### ISSUE THREAD

**Issue by mhemmings** [apex/up#586]
## Prerequisites

* [x] I am running the latest version. (`up upgrade`)
* [x] I searched to see if the issue already exists.
* [x] I inspected the verbose debug output with the `-v, --verbose` flag.
* [x] Are you an Up Pro subscriber?

## Description

Would be great (especially for CI) to be able to separate the build and deploy steps in the Up tool. "Build once, deploy many". Also great for doing things like running e2e tests on the actual thing that's getting deployed, rather than building it again.

Opens up some great possibilities of having an asset that can put in S3 and used at a later date.

It's currently possible to create a zip, so think it's not too much work to have Up take that zip as an input for the deploy step?

**Comment by tj** [comment]
Totally not against the zip thing, but one other feature I wanted was "stage promotion". That way you could deploy `up staging` then if everything looks good, just `up promote staging production` or similar to use the exact same build. 

Having both wouldn't be too bad, maybe a bit redundant but do you think one has a benefit over the other for your use-cases?

**Comment by mhemmings** [comment]
Promotion sounds fantastic.
I see value in both tbh, I think they serve subtly different purposes 🙂

**Comment by agurtovoy** [comment]
FWIW, we implemented build-to-S3-then-deploy=from-zip CI workflow ourselves.

**Comment by mhemmings** [comment]
Using Up? Sounds exactly what we're after! Mind sharing any details?

**Comment by agurtovoy** [comment]
@mhemmings Yep, or I guess sort of: we build the build zips ourselves and put them on S3 (using Gulp), and have a separate `deploy` target that takes the build from S3, unpacks it, and feeds it to `up`. That means that technically we end up with two copies of the "build" on S3, one ours and one `ups`, but we don't mind. S3 is cheap, and It's actually been useful to be able to compare, say, trimmed `node_modules` between our and up`s copies.

Not to downplay your feature request by any means, BTW. If we could prepare the original zip using `up`, we probably would.

**Comment by tj** [comment]
@agurtovoy just jumping in to mention you can do `up build > out.zip`, but we're missing `up --zip out.zip` and/or this promotion type feature

**Comment by davidjnelson** [comment]
a promotion feature would be really great!

**Comment by mhemmings** [comment]
@tj Just a thought, reckon this is possible to do (hack) at the moment by overriding some hooks to not build anything but take a zip as an asset to deploy?

**Comment by tj** [comment]
@mhemmings hmm not that I can think of, the zip portion is somewhat hard-coded in that sense
